### PR TITLE
[Java] Add ability to configure headers

### PR DESCRIPTION
**Pull request by mikaelm12**
Adding ability to configure headers in the Java client.
Fixes: https://github.com/aspnet/SignalR/issues/3064

### CODE REVIEW

File: clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/HubConnectionBuilder.java [hunk 1]
```diff
@@ -43,6 +47,11 @@ public HubConnectionBuilder configureLogging(Logger logger) {
         return this;
     }
 
+    public HubConnectionBuilder withHeaders(Map<String, String> headers) {
```
Comment (mikaelm12) [inline clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/HubConnectionBuilder.java]: Should I add another API to add individual headers?

Comment (mikaelm12) [reply clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/HubConnectionBuilder.java]: `addHeader(String name, String value)`

Comment (anurse) [reply clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/HubConnectionBuilder.java]: Yeah, I think that would be good.

**Comment by mikaelm12** [comment]
>Test?

Yeah, meant to add that a test was still needed in the description

File: clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/HubConnectionBuilder.java [hunk 2]
```diff
@@ -43,6 +47,19 @@ public HubConnectionBuilder configureLogging(Logger logger) {
...
+        this.headers = headers;
+        return this;
+    }
+
+    public HubConnectionBuilder withHeader(String name, String value) {
```
Comment (anurse) [inline clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/HubConnectionBuilder.java]: Better name 👍

File: clients/java/signalr/src/test/java/com/microsoft/aspnet/signalr/HubConnectionTest.java [hunk 3]
```diff
@@ -1053,6 +1055,64 @@ public void accessTokenProviderIsOverriddenFromRedirectNegotiate()
...
+                .build();
+
+        hubConnection.start().get(1000, TimeUnit.MILLISECONDS);
+        assertEquals(HubConnectionState.CONNECTED, hubConnection.getConnectionState());
+        hubConnection.stop();
+        assertEquals("ExampleValue", header.get());
+    }
+
+    @Test
+    public void headersAreSetAndSentThroughBuilder() throws Exception {
```
Comment (anurse) [inline clients/java/signalr/src/test/java/com/microsoft/aspnet/signalr/HubConnectionTest.java]: Can you have a test that calls `withHeader` twice with the same value? I think the best thing to do is to have the second value win (which is what `put` should do by default).

**Comment by mikaelm12** [comment]
This needs to wait on https://github.com/aspnet/SignalR/issues/3097.

**Comment by mikaelm12** [comment]
We're only supporting websockets for the time being

**Comment by anurse** [comment]
> Does Java have equivalent header support for each transport?

Yes, the Java WebSocket library we are using does have header support, so we're good.

File: clients/java/signalr/src/test/java/com/microsoft/aspnet/signalr/HubConnectionTest.java [hunk 4]
```diff
@@ -1053,6 +1055,93 @@ public void accessTokenProviderIsOverriddenFromRedirectNegotiate()
...
+    @Test
+    public void headersAreSetAndSentThroughOptions() throws Exception {
+        AtomicReference<String> header = new AtomicReference<>();
+        TestHttpClient client = new TestHttpClient()
+                .on("POST", "http://example.com/negotiate",
+                        (req) -> {
+                            header.set(req.getHeaders().get("ExampleHeader"));
+                            return CompletableFuture
+                                    .completedFuture(new HttpResponse(200, "", "{\"connectionId\":\"bVOiRPG8-6YiJ6d7ZcTOVQ\",\""
+                                            + "availableTransports\":[{\"transport\":\"WebSockets\",\"transferFormats\":[\"Text\",\"Binary\"]}]}"));
```
Comment (BrennanConroy) [inline clients/java/signalr/src/test/java/com/microsoft/aspnet/signalr/HubConnectionTest.java]: nit: Indenting is a bit aggressive here

File: clients/java/signalr/src/test/java/com/microsoft/aspnet/signalr/HubConnectionTest.java [hunk 5]
```diff
@@ -1053,6 +1055,93 @@ public void accessTokenProviderIsOverriddenFromRedirectNegotiate()
...
+    @Test
+    public void headersAreSetAndSentThroughBuilder() throws Exception {
+        AtomicReference<String> header = new AtomicReference<>();
+        TestHttpClient client = new TestHttpClient()
+                .on("POST", "http://example.com/negotiate",
+                        (req) -> {
+                            header.set(req.getHeaders().get("ExampleHeader"));
+                            return CompletableFuture
+                                    .completedFuture(new HttpResponse(200, "", "{\"connectionId\":\"bVOiRPG8-6YiJ6d7ZcTOVQ\",\""
+                                            + "availableTransports\":[{\"transport\":\"WebSockets\",\"transferFormats\":[\"Text\",\"Binary\"]}]}"));
```
Comment (BrennanConroy) [inline clients/java/signalr/src/test/java/com/microsoft/aspnet/signalr/HubConnectionTest.java]: And here

File: clients/java/signalr/src/test/java/com/microsoft/aspnet/signalr/HubConnectionTest.java [hunk 6]
```diff
@@ -1053,6 +1055,93 @@ public void accessTokenProviderIsOverriddenFromRedirectNegotiate()
...
+    @Test
+    public void sameHeaderSetTwiceGetsOverwritten() throws Exception {
+        AtomicReference<String> header = new AtomicReference<>();
+        TestHttpClient client = new TestHttpClient()
+                .on("POST", "http://example.com/negotiate",
+                        (req) -> {
+                            header.set(req.getHeaders().get("ExampleHeader"));
+                            return CompletableFuture
+                                    .completedFuture(new HttpResponse(200, "", "{\"connectionId\":\"bVOiRPG8-6YiJ6d7ZcTOVQ\",\""
+                                            + "availableTransports\":[{\"transport\":\"WebSockets\",\"transferFormats\":[\"Text\",\"Binary\"]}]}"));
```
Comment (BrennanConroy) [inline clients/java/signalr/src/test/java/com/microsoft/aspnet/signalr/HubConnectionTest.java]: And here

**Pull request merged**